File: 2023-04(백준, 프로그래머스)/키패드 누르기.py
def solution(numbers, hand):
    nums = [[1,4,7],[2,5,8,0],[3,6,9]]
    hands = [[3,1]] + [[j,i] for j in range(3) for i in range(3)]
    l_h = [3,0]
    r_h = [3,2]
    
    result = ''
    for i in numbers:
        if i in nums[0]:
            result += 'L'
            l_h = hands[i]
        elif i in nums[2]:
            result += 'R'
            r_h = hands[i]
        else:
            l_r = abs(l_h[0] - hands[i][0]) + abs(l_h[1] - hands[i][1])
            r_r = abs(r_h[0] - hands[i][0]) + abs(r_h[1] - hands[i][1])
            if l_r == r_r:
                if hand == 'right':
                    result += 'R'
                    r_h = hands[i]
                else:
                    result += 'L'
                    l_h = hands[i]
            else:
                if l_r < r_r:
                    result += 'L'
                    l_h = hands[i]
                else:
                    result += 'R'
                    r_h = hands[i]
    return result

# 뭔가 더 간단한 방법이 있을거같으면서도 생각나지 않았다.
# 너무 노가다 식으로 코드를 작성한 느낌이랄까...

File: JS 코테 연습/키패드 누르기.js
function solution(numbers, hand) {
    const left = [1, 4, 7];
    const mid = { 2: [0, 1], 5: [1, 1], 8: [2, 1], 0: [3, 1] };
    const right = [3, 6, 9];
  
    let now_l = [3, 0];
    let now_r = [3, 2];
  
    let result = numbers.map((n) => {
      if (left.includes(n)) {
        now_l = [Math.floor(n / 3), 0];
        return "L";
      }
      if (right.includes(n)) {
        now_r = [Math.floor(n / 3) - 1, 2];
        return "R";
      }
  
      let now_m = mid[n];
      let ld = Math.abs(now_l[0] - now_m[0]) + Math.abs(now_l[1] - now_m[1]);
      let rd = Math.abs(now_r[0] - now_m[0]) + Math.abs(now_r[1] - now_m[1])
      if (ld < rd) {
        now_l = now_m;
        return "L";
      } else if (ld > rd) {
        now_r = now_m;
        return "R";
      } else {
        if (hand === "left") {
          now_l = now_m;
        } else {
          now_r = now_m;
        }
      }
        return hand[0].toUpperCase();
    });
  
    return result.join('')
  }

solution([1, 3, 4, 5, 8, 2, 1, 4, 5, 9, 5], "right");
